### PR TITLE
feat: improvements to `@scalar/fastify-api-reference`

### LINKED ISSUE #1328: Improvements to `@scalar/fastify-api-reference`
### What happens?

The fastify scalar plugin only uses `fastify-html` to set the correct content type header. The problem lives by `fastify-html` decorating the fastify reply with `.html` which is a common name also used by a lot of other pacakges:

1. @kitajs/fastify-html-plugin
2. @fastify/vite
3. @fastify/react
4. (and others)

Since the main feature of `fastify-html` is layouts (which is not used here), we could simply remove the plugin to avoid such conflict issues. _[The entire plugin source code.](https://github.com/mcollina/fastify-html/blob/main/index.js)_

Also, current setup doesn't install it automatically (its only on peerDeps and devDeps) and npm does not install peer dependencies by default (pnpm with `auto-install-deps` does, but not everyone uses it). So if the plugin will be keept, it should be moved to the package's dependencies.

### What did you expect to happen?

above :)

### How can we reproduce the issue?

above :)

### Swagger/OpenAPI Example

_No response_

### PR DESCRIPTION
**Pull request by arthurfiorette**
closes #1328.

### CODE REVIEW

**Comment by marclave** [comment]
lets go @arthurfiorette ! great to see you here :)

will review in a bit im just deep in some test writing :)

**Comment by arthurfiorette** [comment]
Hey! It's me again haha

No problem, take your time :).

I couldn't run the tests on my machine, so I'll wait for the workflow approval to fix any pending issues...

**Comment by marclave** [comment]
> Hey! It's me again haha
> 
> No problem, take your time :).
> 
> I couldn't run the tests on my machine, so I'll wait for the workflow approval to fix any pending issues...

i'll run those now! what errors did you hit running locally? 

:)

**Comment by arthurfiorette** [comment]
> i'll run those now! what errors did you hit running locally?

Oh it was my fault, I needed to run `pnpm build` at the project root. It now pass just fine!

**Comment by arthurfiorette** [comment]
Whoops, By updating fastify-api-reference fastify version, I also needed to update other fastify dependencies. It should now be fixed (tested locally 🫡)

**Comment by arthurfiorette** [comment]
wait... Somehow my LSP plugin updated a lot of other dependencies too hahaha should I rollback it?

**Comment by marclave** [comment]
> wait... Somehow my LSP plugin updated a lot of other dependencies too hahaha should I rollback it? (Everything seems to be working just fine)

ah yeah probably rollback just incase :)

**Comment by arthurfiorette** [comment]
Hmm, it actually wasn't me. Something is being ran at pre-commit... anyways, i manually skipped the pre-commit hook and reverted it. 

Ready for review :)

**Comment by arthurfiorette** [comment]
> I also saw that `fastify-plugin` isn't used as well... Should it be removed from `package.json` too?

Even if this plugin gets encapsulated or not, `fastify-plugin` should be used to define all metadata required by fastify, such as `fastify.display-name` which later allows to check if this plugin was registered or not with `fastify.hasPlugin('@scalar/fastify-api-reference')`...

**Comment by arthurfiorette** [comment]
I didn't use encapsulate here:

https://github.com/scalar/scalar/blob/8a25c1182fefb85fa8caa9255c509cb2db339015/packages/fastify-api-reference/src/fastifyApiReference.ts#L241-L243

Because the `fastify-plugin` is not used anymore, since it has no reason to... 

---

But if you guys want to keep `fastify-html`, as stated [here](https://github.com/scalar/scalar/issues/1328#issuecomment-2024348648) and [theoretically has no need to change](https://github.com/scalar/scalar/issues/1328#issuecomment-2025583376). I could add it back and also add `encapsulate: true` to ensure it keeps encapsulated and dont mess up with fastify's "globals".

`encapsulate: true` has the same effect of not using `fastify-plugin` at all, but using it keeps other advantages of `fastify-plugin` such as plugin name, as previously stated.

**Comment by arthurfiorette** [comment]
Sorry for asking @marclave, but you review this pr 😅😁?

**Comment by marclave** [comment]
> Sorry for asking @marclave, but you review this pr 😅😁?

ah! sorry this slipped through, I will do a proper review tomorrow : ) just trying to get some other things in today ✨ 

really sorry about that

@mcollina thoughts on these comments above ✨

File: packages/fastify-api-reference/package.json
Comment (mcollina) [inline packages/fastify-api-reference/package.json:61]: I would actually recommend removing peerDependencies. They just cause problems.

**Review by mcollina — APPROVED** [review]
lgtm

File: .changeset/hot-plums-fail.md
Comment (mcollina) [inline .changeset/hot-plums-fail.md:5]: this is incorrect. There are no conflicts.

**Comment by mcollina** [comment]
Note that this does not technically fix https://github.com/scalar/scalar/issues/1328, because that's not an actual problem.

**Comment by arthurfiorette** [comment]
> Hmm... Interesting, seems that I was in fact wrong... [Its a known limitation of fastify vite](https://fastify-vite.dev/guide/known-limitations#known-limitations). 

https://github.com/scalar/scalar/issues/1328#issuecomment-2025604485

File: .changeset/hot-plums-fail.md
Comment (arthurfiorette) [reply .changeset/hot-plums-fail.md:5]: Yes, ofc! I'll change these messages. 

https://github.com/scalar/scalar/issues/1328#issuecomment-2025604485

File: packages/fastify-api-reference/package.json
Comment (arthurfiorette) [reply packages/fastify-api-reference/package.json:61]: I thought it was just with me XD

**Comment by arthurfiorette** [comment]
@marclave ready (again) for review/merge!

**Comment by marclave** [comment]
appreciate everyones work and reviews on this!

@amritk and I are gunna review this now : )

**Comment by marclave** [comment]
@arthurfiorette looks like we need to fix up the lock file, you can delete it and do a fresh `pnpm install` at root of the scalar repo ✨

**Comment by marclave** [comment]
hey @arthurfiorette we are getting some build + type errors  in CI 

```
Error: src/fastifyApiReference.ts(165,7): error TS2742: The inferred type of 'fastifyApiReference' cannot be named without a reference to '.pnpm/fastify@4.26.2/node_modules/fastify'. This is likely not portable. A type annotation is necessary.
```

**Comment by marclave** [comment]
@arthurfiorette I just had to add this : )

```
diff --git a/packages/fastify-api-reference/src/fastifyApiReference.ts b/packages/fastify-api-reference/src/fastifyApiReference.ts
index 78ec732e..6a3baa72 100644
--- a/packages/fastify-api-reference/src/fastifyApiReference.ts
+++ b/packages/fastify-api-reference/src/fastifyApiReference.ts
@@ -1,4 +1,10 @@
 import type { ReferenceConfiguration } from '@scalar/api-reference'
+import {
+  type FastifyBaseLogger,
+  type FastifyPluginCallback,
+  type FastifyTypeProviderDefault,
+  type RawServerDefault,
+} from 'fastify'
 import fp from 'fastify-plugin'
 
 import { getJavaScriptFile } from './utils'
@@ -162,7 +168,12 @@ export function htmlDocument(options: FastifyApiReferenceOptions) {
 `
 }
 
-const fastifyApiReference = fp<FastifyApiReferenceOptions>(
+const fastifyApiReference: FastifyPluginCallback<
+  FastifyApiReferenceOptions,
+  RawServerDefault,
+  FastifyTypeProviderDefault,
+  FastifyBaseLogger
+> = fp<FastifyApiReferenceOptions>(
   async (fastify, options) => {
     let { configuration } = options
     const hasSwaggerPlugin = fastify.hasPlugin('@fastify/swagger')
     ```

**Comment by arthurfiorette** [comment]
@marclave anything else?

**Comment by marclave** [comment]
@arthurfiorette okay tests are passing! last thing could you rebase, supposedly there are conflicts : )

**Review by marclave — APPROVED** [review]
✨ awesome work @arthurfiorette !!!

**Comment by marclave** [comment]
oh wait nvm it was just github glitching out! let's get this in ✨

**Pull request merged**
